Main.hs

```diff
@@ -2,28 +2,130 @@ module Main where
 
 import Control.Monad (unless)
 import Text.Printf (printf)
+import Language.Haskell.TH (Exp, safe)
+import Data.Either (isLeft, fromRight)
+import Control.Arrow (ArrowChoice(right))
 
-data Expr
+data Expr = Numb Double
+      | SqrtExpr Expr
+      | BinOpAdd Expr Expr
+      | BinOpSub Expr Expr
+      | BinOpMul Expr Expr
+      | BinOpDiv Expr Expr
+      | BinOpPow Expr Expr
 
 instance Show Expr where
-  show = undefined
+  show (Numb a) = printf "%s" (show a)
+  show (SqrtExpr a) = printf "sqrt %s" (show a)
+  show (BinOpAdd a b) = printf "(%s + %s)" (show a) (show b)
+  show (BinOpSub a b) = printf "(%s - %s)" (show a) (show b)
+  show (BinOpMul a b) = printf "(%s * %s)" (show a) (show b)
+  show (BinOpDiv a b) = printf "(%s / %s)" (show a) (show b)
+  show (BinOpPow a b) = printf "(%s ^ %s)" (show a) (show b)
 
 instance Eq Expr where
-  (==) = undefined
+  Numb a == Numb b = a == b
+  SqrtExpr a == SqrtExpr b = a == b
+  BinOpAdd a b == BinOpAdd c d = (a == c) && (b == d)
+  BinOpSub a b == BinOpSub c d = (a == c) && (b == d)
+  BinOpMul a b == BinOpMul c d = (a == c) && (b == d)
+  BinOpDiv a b == BinOpDiv c d = (a == c) && (b == d)
+  BinOpPow a b == BinOpPow c d = (a == c) && (b == d)
+  _ == _ = False
 
-data Error
+data Error = DivByZero String | SqrtNeg String
 
 instance Show Error where
-  show = undefined
+  show (DivByZero s) = printf "%s" (show s)
+  show (SqrtNeg s) = show s
 
 instance Eq Error where
-  (==) = undefined
+  DivByZero s1 == DivByZero s2 = s1 == s2
+  SqrtNeg s1 == SqrtNeg s2 = s1 == s2
+  _ == _ = False
 
 eval :: Expr -> Either Error Double
-eval = undefined
+eval (Numb a) = Right a
+eval (SqrtExpr a)
+    | isLeft aAfterEval = aAfterEval
+    | aVal < 0 = Left (SqrtNeg "Arg for sqrt cannot be negative")
```

Comment:
The information "Arg for sqrt cannot be negative" is already encoded in the constructor of this error, no need to duplicate it. Wrapping the sub-expression, which caused the error, is more informative. You can make use of this string in the `Show` instance of the error.